### PR TITLE
server: do not crash if diagnostics loop panics

### PR DESCRIPTION
**Pull request by dt**
Losing diagnostics is a bummer for us (cockroach labs), but shouldn't cause a disruption for a user (unless they've overridden the diagnostics endpoint and are running an on-prem collector).

The one caveat is that the diagnostics loop is usually charged with periodically clearing the stmt stats stats that could otherwise grow unbounded -- so if we allow that loop to crash, we need a fallback that ensures they still get discarded (error counts counted by the fixed set of error/feature codes and are thus not unbounded).

### CODE REVIEW

**Comment by dt** [comment]
I'm not dead-set on this -- we shouldn't be panicking in the loop in the first place, and I'd rather just fix those. But if one does slip though, it's a pretty big bummer to have the user's nodes crash and become unavailable when there isn't a core correctness concern or anything, so I went ahead and whipped this up.

**Comment by bdarnell** [comment]
:lgtm:

Have we seen panics here or is it just a precaution?

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24511#-:-L9LaNwDDO-ujdytyXnI:bb0xc2i)*

**Comment by dt** [comment]
@bdarnell unfortunately https://github.com/cockroachdb/cockroach/issues/24499

File: pkg/server/updates.go [hunk 1]
```diff
@@ -97,10 +103,40 @@ type versionInfo struct {
 	Details string `json:"details"`
 }
 
+// PeriodicallyClearStmtStats is backstop -- generally the diagnostics reporting
```
Comment (knz) [inline pkg/server/updates.go]: I'm really sorry but I don't understand this description comment.
When I start reading this I want to know:
- what this function does on the first line
- what is the mechanism by which it guarantees that no stats are lost via a reset if reporting does work (incidentally: I do not see any such mechanism in the implementation below)
- not too many sub-clauses in each sentence. The last sentence is particularly hard to parse.

Also incidentally I do not understand "it drives that reset"

**Comment by dt** [comment]
yikes, yeah, that comment was a mess @BramGruneir @knz -- reworded a bit now.

**Comment by BramGruneir** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24511#-:-L9MWFzI5PB3ZKmvgl_J:bnfp4nl)*

**Comment by knz** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r1, 1 of 2 files at r2, 2 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24511#-:-L9Ml18F0Vqoiy_dE47G:bnfp4nl)*

File: pkg/sql/app_stats.go [hunk 2]
```diff
@@ -349,6 +353,14 @@ func (e *Executor) ResetStatementStats(ctx context.Context) {
 	e.sqlStats.resetStats(ctx)
 }
 
+// LasStatementStatReset returns the time the stmt stats were last rest.
+func (e *Executor) LasStatementStatReset() time.Time {
+	e.sqlStats.Lock()
```
Comment (andreimatei) [inline pkg/sql/app_stats.go]: I don't think this does what you think it does... 
The `sqlStats` you want I believe are these: https://github.com/cockroachdb/cockroach/blob/58335bcaa3faa1ff27981c22e27b998125470983/pkg/sql/conn_executor.go#L229

I see there is a test change in this commit; does that prove that I'm confused?

Generally, the `Executor` is mostly unused code, so it shouldn't be touched.

Comment (andreimatei) [reply pkg/sql/app_stats.go]: Addressed in #24941

Comment (dt) [reply pkg/sql/app_stats.go]: We should really nuke Executor if it is dead code.

Comment (andreimatei) [reply pkg/sql/app_stats.go]: trying, trust me
